autorest/azure/auth/auth.go

```diff
@@ -0,0 +1,197 @@
+package auth
+
+// Copyright 2017 Microsoft Corporation
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+
+import (
+	"crypto/rsa"
+	"crypto/x509"
+	"fmt"
+	"io/ioutil"
+	"os"
+
+	"github.com/Azure/go-autorest/autorest"
+	"github.com/Azure/go-autorest/autorest/adal"
+	"github.com/Azure/go-autorest/autorest/azure"
+	"golang.org/x/crypto/pkcs12"
+)
+
+// GetAuthorizerWithDefaults tries to get an authorizer using the following methods:
+// 1. Client credentials
+// 2. Client certificate
+// 3. MSI
+func GetAuthorizerWithDefaults() (*autorest.BearerAuthorizer, error) {
+
+	options := NewAuthorizerOptionsFromEnvironment()
+
+	//1.Client Credentials
+	authorizer, err := GetAuthorizerFromClientCredentials(options)
+
+	//2. Client Certificate
+	if err != nil {
+		authorizer, err = GetAuthorizerFromClientCertificate(options)
+	}
+
+	//3. MSI
+	if err != nil {
+		authorizer, err = GetAuthorizerFromMSI(options.Environment)
+	}
+
+	return authorizer, nil
+}
+
+// GetAuthorizerFromMSI gets an authorizer from MSI. Note that this will only work when running in an Azure environment
+func GetAuthorizerFromMSI(env azure.Environment) (*autorest.BearerAuthorizer, error) {
+	msiEndpoint, err := adal.GetMSIVMEndpoint()
+
+	if err != nil {
+		return nil, err
+	}
+
+	spToken, err := adal.NewServicePrincipalTokenFromMSI(msiEndpoint, env.ResourceManagerEndpoint)
+
+	if err != nil {
+		return nil, fmt.Errorf("failed to get oauth token from MSI: %v", err)
+	}
+
+	return autorest.NewBearerAuthorizer(spToken), nil
+}
+
+// GetAuthorizerFromClientCredentials gets an authorizer from client credentials.
+func GetAuthorizerFromClientCredentials(options AuthorizerOptions) (*autorest.BearerAuthorizer, error) {
+	oauthConfig, err := adal.NewOAuthConfig(options.Environment.ActiveDirectoryEndpoint, options.TenantID)
+	if err != nil {
+		return nil, err
+	}
+
+	spToken, err := adal.NewServicePrincipalToken(*oauthConfig, options.ClientID, options.ClientSecret, options.Environment.ResourceManagerEndpoint)
```

Comment:
can we factor the endpoint into the authorizer options so another endpoint can be chosen?